src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,231 @@
+#include <IOKit/IOReturn.h>
+#include <IOKit/hidsystem/IOHIDLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <CoreFoundation/CoreFoundation.h>
+#undef Boolean
+#endif
+
+enum {
+    kNumLock = 0x10,
+    kCapsLock = 0x02,
+};
+
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+static _XDisplay *(*_XOpenDisplay)(void *);
+static int (*_XCloseDisplay)(_XDisplay *);
+static _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+
+/// A helper function to be used to load the libX11 and required functions
+static bool os_xlib_init(const char **const err)
+{
+  if (xlib_opened) {
+    return true;
+  }
+  const int udl_ret = uv_dlopen(XLIB_NAME, &xlib);
+  if (udl_ret != 0) {
+    *err = "Failed to open library libX11";
+  } else {
+    xlib_opened = true;
+  }
+
+  if ((uv_dlsym(&xlib, "XOpenDisplay", (void **)&_XOpenDisplay)
+       || uv_dlsym(&xlib, "XCloseDisplay", (void **)&_XCloseDisplay)
+       || uv_dlsym(&xlib, "XkbGetState", (void **)&_XkbGetState)) != 0) {
+    *err = "Unable to get the functions of X11";
+    return false;
+  }
+  return xlib_opened;
+}
+
+/// Get information about Num/Caps Lock state in linux
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmas
+///
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+#define CLOSE_DISPLAY(display) \
+  do { \
+    const int e = _XCloseDisplay(display); \
+    if (e) { \
+      *err = "XCloseDisplay unable to close the display"; \
+      return false; \
+    } \
+  } while (0)
+  *mods = 0;
+  *err = NULL;
+  if (!os_xlib_init(err)) {
+    return false;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    *err = "Couldn't open the display";
+    return false;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    *err = "Non-zero status returned by XkbGetState";
+    CLOSE_DISPLAY(display);
+  }
+  *mods = xkb_state.data[kLockedModsOffset];
+
+  CLOSE_DISPLAY(display);
+  return true;
+#undef CLOSE_DISPLAY
+}
+
+#elif defined(WIN32)
+/// Get information about Num/Caps Lock state in Windows
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmas
+///
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return (GetKeyState(VK_CAPITAL) & 0x0001) != 0;
```

Comment:
Function does not do anything useful due to this. BTW, is not there a possibility to get scroll lock state there as well? It is better to omit relevant key in OS X only rather then everywhere.

Reply:
Okk, here it is possible to get the scrollock, and may be I forgot to remove this line, thanks for the review :smiley: